chore: add global caching to rbac

ammario opened the pull request:
I'm assuming all calls to Authorize are pure in this implementation.

ammario committed:
chore: use global cache for RBAC

ammario committed:
fix build errors

ammario committed:
rm coderd.test

ammario committed:
Rework cache bench

ammario commented:
thoughts on this approach @Emyrk?

Emyrk commented on coderd/rbac/authz.go:
```diff
@@ -617,94 +626,37 @@ type cachedCalls struct {
+func (c *authCache) Authorize(ctx context.Context, subject Subject, action Action, object Object) error {
+	var authorizeCacheKey strings.Builder
+	enc := json.NewEncoder(&authorizeCacheKey)
+	_ = enc.Encode(subject)
+	_ = enc.Encode(action)
+	_ = enc.Encode(object)
```
This is unfortunate. The json payload is quite large, and that being a key would be even more verbose.

It would be better to "hash" the fields somehow to reduce this.

I intentionally made the file [`astvalue.go`](https://github.com/coder/coder/blob/6ed35b406a7a0a3ec877ed41d11da904d243ad09/coderd/rbac/astvalue.go#L11-L33) to avoid json.Marshalling this data when going to the rego evaluator.

----

We can use the role name for the hash and ignore the permissions. Whatever "hash" function we create will be faster than a json.Marshal I bet.

Emyrk reviewed (commented):
I'm not opposed to this. I'm curious what kind of savings this would yield. We should not use the full json payload as a key though.

---

The caching per request was intentional as a single request often does the same rbac check a lot. Example is fetching a workspace. You have to fetch the workspace build and the provisioner job, which all do the same rbac check on the workspace.

Going global means that we might lose this savings per request if we get more than the 4096 requests in the same time of a single request. Probably unlikely? So it's probably ok?

Just something to think about. Since the per request cache will guarantee duplicate hits. Might be overkill to have both a per request cache and global cache.

ammario commented:
@Emyrk — I have a benchmark that is just calling `/users/me` and the cache is taking us from 3.3ms/req to 0.7ms/req, so quite significant. I agree that hashing the key is a good idea, and since that would reduce storage overhead, I think we can just make the cache really big, like 64k entries.

Emyrk commented:
From a correctness POV, as long as our unique key contains all the information required to match a unique RBAC check, then it will be correct in all future calls too.

So this can work for sure :+1:.

ammario commented:
@Emyrk very happy to hear that :)

Emyrk commented:
@ammario If we go to 64k, then I have no issues with dropping the request cache. Can I write a hash function and commit it to this branch?

ammario commented:
@Emyrk go for it

Emyrk committed:
chore: Add ability to hash an authorizationc all to a unqiue hash key.

This is useful for caching purposes.
Order matters for the hash key for slices, so a slice with
the same elements in a different order is a different hash key.
Athlough they are functionally equivalent, our slices have deterministic
orders in the database, so our subjects/objects will match.
There is no need to spend the time or memory to sort these lists.

Emyrk committed:
Remove import

ammario committed:
Do not record authorization calls in benchmarks

ammario committed:
ignore .test binaries

ammario committed:
pass lint

ammario committed:
Use JSON again

ammario committed:
Merge remote-tracking branch 'origin/main' into rbac-global-cache

Emyrk reviewed (approved):
:+1:
